examples/post-sync/controller.go

```diff
@@ -0,0 +1,312 @@
+// limitations under the License.
+
+package main
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/hex"
+	"fmt"
+	"strings"
+	"sync"
+
+	"github.com/go-logr/logr"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"kpt.dev/configsync/pkg/api/configsync/v1beta1"
+)
+
+// Sync kinds
+const (
+	RootSyncKind = "RootSync"
+	RepoSyncKind = "RepoSync"
+)
+
+// SyncID uniquely identifies a sync resource
+type SyncID struct {
+	Name      string
+	Kind      string
+	Namespace string
+}
+
+// StatusTracker tracks which errors have been logged
+type StatusTracker struct {
+	mu   sync.Mutex
+	seen map[string]struct{} // Using a single map with hashed keys
+}
+
+// NewStatusTracker creates a new StatusTracker instance
+func NewStatusTracker() *StatusTracker {
+	return &StatusTracker{
+		seen: make(map[string]struct{}),
+	}
+}
+
+// generateKey creates a unique key for a sync resource, commit, and error message
+func generateKey(syncID SyncID, commit, message string) string {
+	// Include SyncID information in the key
+	key := fmt.Sprintf("%s:%s:%s:%s:%s",
+		syncID.Name,
+		syncID.Kind,
+		syncID.Namespace,
+		commit,
+		message)
+
+	// Hash the key to save memory
+	hash := sha256.Sum256([]byte(key))
+	return hex.EncodeToString(hash[:])
+}
+
+// IsLogged checks if an error has been logged
+func (s *StatusTracker) IsLogged(syncID SyncID, commit, message string) bool {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	key := generateKey(syncID, commit, message)
+	_, logged := s.seen[key]
+	return logged
+}
+
+// MarkLogged marks an error as logged
+func (s *StatusTracker) MarkLogged(syncID SyncID, commit, message string) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	key := generateKey(syncID, commit, message)
+	s.seen[key] = struct{}{}
+}
+
+// SyncStatusController reconciles RootSync and RepoSync resources
+type SyncStatusController struct {
+	client        client.Client
+	log           logr.Logger
+	statusTracker *StatusTracker
+	syncKind      string // Explicit field to indicate what kind of resource we're watching
+}
+
+// NewSyncStatusController creates a new controller instance
+func NewSyncStatusController(client client.Client, log logr.Logger, statusTracker *StatusTracker, syncKind string) *SyncStatusController {
+	// Validate sync kind
+	if syncKind != RootSyncKind && syncKind != RepoSyncKind {
+		panic(fmt.Sprintf("invalid sync kind: %s, must be either %s or %s", syncKind, RootSyncKind, RepoSyncKind))
+	}
+	return &SyncStatusController{
+		client:        client,
+		log:           log,
+		statusTracker: statusTracker,
+		syncKind:      syncKind,
+	}
+}
+
+// Reconcile handles RootSync and RepoSync resources
+func (c *SyncStatusController) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	var syncID SyncID
+	var status v1beta1.Status
+	var generation int64
+	var observedGeneration int64
+	var errs, commit string
```

Comment:
nit: errs -> errMessage, like the other one down below.

Reply:
done

Reply:
Rephrased to conditionErrorMessage and statusErrorMessage for readability